### ISSUE TITLE
useQuery typings fails to require required variables from TypedDocumentNode, causing runtime error

### ISSUE THREAD

**Issue by irond13** [apollographql/apollo-client#10857]
### Issue Description

Hi!

The current typings for `useQuery` are as follows:
```
export declare function useQuery<TData = any, TVariables extends OperationVariables = OperationVariables>(query: DocumentNode | TypedDocumentNode<TData, TVariables>, options?: QueryHookOptions<TData, TVariables>): QueryResult<TData, TVariables>;
```

Note that `options` is always optional. This means it is possible to pass a `TypedDocumentNode` which defines non-optional variables to `useQuery` without actually passing any of the required variables.

This will result in runtime error, thrown by the GraphQL evaluator, which would expect said variables.

### Link to Reproduction

https://codesandbox.io/p/sandbox/dazzling-borg-o6fr3u?welcome=true

### Reproduction Steps

See file `useQueryTest.tsx` in the linked sandbox. The expectation is that the uncommented `useQuery` should fail to compile, which it doesn't.

**Comment by nandin-borjigin** [comment]
Please also correct the typing on `options.variables`. It's optional as well so even if the `options` was required, we can still incorrectly call it as 
```
useQuery(
  someTypedQueryDocument, 
  {/* variables missing while it's required for someTypedQueryDocument */}
)
```

**Comment by phryneas** [comment]
This requires a bunch of bigger changes (as it would apply to all hooks).
You can follow the process over in #11241

**Comment by tannerbaum** [comment]
Forgive me if I am mistaken, is this the same thing I am experiencing with my generated types through graphql-codegen and @apollo/client which worked fine until a round of dependency updates? 

`Argument of type 'QueryHookOptions<PageContentQuery, Exact<{ slug: string; locale: Gcms_Locale; }>>' is not assignable to parameter of type 'QueryHookOptions<any, OperationVariables>'.`

**Comment by phryneas** [comment]
@tannerbaum no, that's a different issue. I'd be grateful if you could open a new issue and provide a small reproduction that shows the issue so we can investigate it.

**Comment by phryneas** [comment]
@jigar-riyalto it's an unrelated issue - could you please open a new issue and also show the code that is causing this type error?

**Comment by phryneas** [comment]
This will be solved with https://github.com/apollographql/apollo-client/pull/12559, which will be released as part of 4.0

**Comment by jerelmiller** [comment]
Hey all 👋 

We just released a change in [`4.0.0-alpha.10`](https://github.com/apollographql/apollo-client/releases/tag/v4.0.0-alpha.10) that provides a lot better type safety around required variables throughout the client. We'd appreciate you trying this out and providing any feedback you can!

Since this change has been merged and published, I'm going to go ahead and close this issue. Please open a new issue if you're experiencing issues with the alpha version on this change. Thanks!

**Comment by irond13** [comment]
Thanks so much @jerelmiller - really appreciate that! I’m off atm, but will check when I’m back.
